azurelinuxagent/common/utils/archive.py

```diff
@@ -41,71 +41,89 @@
 MAX_ARCHIVED_STATES = 50
 
 CACHE_PATTERNS = [
-    re.compile("^(.*)\.(\d+)\.(agentsManifest)$", re.IGNORECASE), # pylint: disable=W1401
-    re.compile("^(.*)\.(\d+)\.(manifest\.xml)$", re.IGNORECASE), # pylint: disable=W1401
-    re.compile("^(.*)\.(\d+)\.(xml)$", re.IGNORECASE) # pylint: disable=W1401
+    re.compile(r"^(.*)\.(\d+)\.(agentsManifest)$", re.IGNORECASE),
+    re.compile(r"^(.*)\.(\d+)\.(manifest\.xml)$", re.IGNORECASE),
+    re.compile(r"^(.*)\.(\d+)\.(xml)$", re.IGNORECASE)
 ]
 
+GOAL_STATE_PATTERN = re.compile(r"^(.*)GoalState\.(\d+)\.xml$", re.IGNORECASE)
+
 # 2018-04-06T08:21:37.142697
 # 2018-04-06T08:21:37.142697.zip
-ARCHIVE_PATTERNS_DIRECTORY = re.compile('^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d+$') # pylint: disable=W1401
-ARCHIVE_PATTERNS_ZIP       = re.compile('^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d+\.zip$') # pylint: disable=W1401
+ARCHIVE_PATTERNS_DIRECTORY = re.compile(r"^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d+$")
+ARCHIVE_PATTERNS_ZIP = re.compile(r"^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d+\.zip$")
 
 
 class StateFlusher(object):
     def __init__(self, lib_dir):
         self._source = lib_dir
 
-        d = os.path.join(self._source, ARCHIVE_DIRECTORY_NAME) # pylint: disable=C0103
-        if not os.path.exists(d):
+        directory = os.path.join(self._source, ARCHIVE_DIRECTORY_NAME)
+        if not os.path.exists(directory):
             try:
-                fileutil.mkdir(d)
-            except OSError as e: # pylint: disable=C0103
+                fileutil.mkdir(directory)
+            except OSError as e:  # pylint: disable=C0103
                 if e.errno != errno.EEXIST:
                     logger.error("{0} : {1}", self._source, e.strerror)
 
-    def flush(self, timestamp):
+    def flush(self):
         files = self._get_files_to_archive()
-        if len(files) == 0: # pylint: disable=len-as-condition
+        if len(files) == 0:  # pylint: disable=len-as-condition
             return
 
-        if self._mkdir(timestamp):
-            self._archive(files, timestamp)
+        goal_state_timestamp = self._get_latest_timestamp(files)
+        if goal_state_timestamp and self._mkdir(goal_state_timestamp):
+            self._archive(files, goal_state_timestamp)
         else:
             self._purge(files)
 
     def history_dir(self, timestamp):
         return os.path.join(self._source, ARCHIVE_DIRECTORY_NAME, timestamp.isoformat())
 
+    @staticmethod
+    def _get_latest_timestamp(files):
+        # Get the most recently modified GoalState.*.xml (if there are more than one) and use that timestamp for the archive name.
+        latest_timestamp_ms = None
+        for current_file in files:
+            match = GOAL_STATE_PATTERN.match(current_file)
```

Comment:
Just brainstorming here, how about getting the incarnation of the goalstate (goalstate.incarnation) and checking if that `GoalState.<incarnation>.xml` file already exists and using the timestamp of that rather than all files that match that regex?

Reply:
The problem is `self._goal_state` can be None and relying on the highest incarnation number on file is not deterministic either, since in the VM migrate scenario we can start getting incarnation 0 even though the previous one was N.

Reply:
From the code, we set the `self._goal_state` right before calling the save function 

```python
            if new_goal_state is not None:
                self._goal_state = new_goal_state
                self._save_goal_state()
```

Essentially all I'm saying is on service restart, if we re-run the same goalstate the incarnation would be the same. So we would already have the `GoalState.<incarnation>.xml` file present on filesystem. Wouldn't it be better to just check the timestamp for that file? That would also take care of the other comment I had earlier where there might be a chance of having multiple `Goalstate.*.xml` files?

Reply:
Which <incarnation> are you referring to? `self.goal_state` can be None, and current goal state is the one we just fetched so that file is not saved to disk yet.

Reply:
let's sync offline about this, it would just be faster that way